DeviceFor instantiates contiguous iterator trait for tabulate output iterators

## Summary

A RAPIDS cuDF build failed after #8565 with a `tuple<..., void>` compilation error while compiling `cpp/src/join/sort_merge_join.cu`. The failure appears to be caused by `cub::DeviceFor::__for_each_n` evaluating `thrust::is_contiguous_iterator_v<RandomAccessIteratorT>` even when vectorization is disabled.

A fix is proposed in #8778.

## Failure

From https://github.com/NVIDIA/cccl/actions/runs/25033689578/job/73320636248:

```text
/home/coder/cudf/cpp/src/join/sort_merge_join.cu:338:20:   required from 'std::pair<std::unique_ptr<rmm::device_uvector<int> >, std::unique_ptr<rmm::device_uvector<int> > > cudf::_GLOBAL__N__e951d515_18_sort_merge_join_cu_7bed6c89::merge<LargerIterator, SmallerIterator>::inner(rmm::cuda_stream_view, rmm::device_async_resource_ref) [with LargerIterator = const int*; SmallerIterator = const int*; rmm::device_async_resource_ref = cuda::mr::__4::__version_bump_ver4_::resource_ref<cuda::mr::__4::device_accessible>]'
/home/coder/cudf/cpp/src/join/sort_merge_join.cu:833:73:   required from 'cudf::detail::sort_merge_join::inner_join(const cudf::table_view&, cudf::sorted, rmm::cuda_stream_view, rmm::device_async_resource_ref) const::<lambda(auto:235&)> [with auto:235 = cudf::_GLOBAL__N__e951d515_18_sort_merge_join_cu_7bed6c89::merge<const int*, const int*>]'
/home/coder/cudf/cpp/src/join/sort_merge_join.cu:774:10:   required from 'auto cudf::detail::sort_merge_join::invoke_merge(const preprocessed_table&, cudf::table_view, cudf::table_view, MergeOperation&&, rmm::cuda_stream_view) const [with MergeOperation = cudf::detail::sort_merge_join::inner_join(const cudf::table_view&, cudf::sorted, rmm::cuda_stream_view, rmm::device_async_resource_ref) const::<lambda(auto:235&)>]'
/home/coder/cudf/cpp/src/join/sort_merge_join.cu:828:26:   required from here
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/sfinae_helpers.h:46:5: error: 'cuda::std::__4::_TestSynthesizeAssignment<_Tp>::__dummy' has incomplete type
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/sfinae_helpers.h:46:5: error: invalid use of 'void'
```

The diagnostic continues with the `void` tuple instantiation:

```text
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/tuple_leaf.h: In instantiation of 'class cuda::std::__4::__tuple_leaf<1, void, cuda::std::__4::__tuple_leaf_specialization::__default>':
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/tuple_leaf.h:376:8:   required from 'struct cuda::std::__4::__tuple_impl<cuda::std::__4::__tuple_indices<0, 1>, cudf::detail::row::rhs_index_type, void>'
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/tuple.h:52:8:   required from 'class cuda::std::__4::tuple<cudf::detail::row::rhs_index_type, void>'
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__type_traits/common_reference.h:75:70:   required by substitution of 'template<class _Ap, class _Bp> struct cuda::std::__4::__common_ref<_Ap&, _Bp&, typename cuda::std::__4::enable_if<is_reference_v<decltype ((false ?  declval<typename decltype (__apply_cv<_From>)::__call<_Yp>& (&)(), void>()() : declval<typename decltype (__apply_cv<_Yp>)::__call<_Xp>& (&)(), void>()()))>, void>::type> [with _Ap = thrust::detail::tuple_of_iterator_references<cudf::detail::row::rhs_index_type, thrust::detail::tabulate_output_iterator_proxy<...> >; _Bp = cuda::std::__4::tuple<cudf::detail::row::rhs_index_type, void>]'
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__type_traits/common_reference.h:198:8:   required from 'struct cuda::std::__4::common_reference<thrust::detail::tuple_of_iterator_references<cudf::detail::row::rhs_index_type, thrust::detail::tabulate_output_iterator_proxy<...> >&&, cuda::std::__4::tuple<cudf::detail::row::rhs_index_type, void>&>'
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/thrust/../../../thrust/thrust/type_traits/is_contiguous_iterator.h:100:86:   required from 'constexpr const bool thrust::detail::is_contiguous_iterator_impl_v<thrust::zip_iterator<cuda::std::__4::tuple<cuda::__4::transform_iterator<...>, thrust::tabulate_output_iterator<...> > > >'
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/thrust/../../../thrust/thrust/type_traits/is_contiguous_iterator.h:133:36:   required from 'constexpr const bool thrust::is_contiguous_iterator_v<thrust::zip_iterator<cuda::std::__4::tuple<cuda::__4::transform_iterator<...>, thrust::tabulate_output_iterator<...> > > >'
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/cub/../../../cub/cub/device/device_for.cuh:125:48:   [ skipping 6 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
```

And later:

```text
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/tuple_leaf.h:97:5: error: 'cuda::std::__4::__tuple_leaf<_Ip, _Hp, cuda::std::__4::__tuple_leaf_specialization::__default>::__value_' has incomplete type
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/tuple_leaf.h:97:5: error: invalid use of 'void'
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/tuple_leaf.h:177:70: error: forming reference to void
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/tuple_leaf.h:181:76: error: forming reference to void
/home/coder/cudf/cpp/build/conda/cuda-13.0/release/_deps/cccl-src/lib/cmake/libcudacxx/../../../libcudacxx/include/cuda/std/__tuple_dir/tuple.h: In instantiation of 'class cuda::std::__4::tuple<cudf::detail::row::rhs_index_type, void>':
```

## Root cause

#8565 refactored `DeviceFor` to route `ForEachN` through `__for_each_n`. In that function, the vectorization check is written as a single `if constexpr` condition:

```cpp
allow_vectorization &&
THRUST_NS_QUALIFIER::is_contiguous_iterator_v<RandomAccessIteratorT>
```

For `ForEachN`, `allow_vectorization` is currently false. However, the compiler still has to form the full constant expression and therefore instantiates `thrust::is_contiguous_iterator_v<RandomAccessIteratorT>`.

That trait is not safe to instantiate for the iterator type generated by `thrust::lower_bound` in the failing cuDF path: a `thrust::zip_iterator` containing a tabulate output iterator. The output iterator has `value_type = void`, so the trait path attempts to form `cuda::std::tuple<..., void>`, which is ill-formed.

## Expected behavior

If vectorization is disabled, `DeviceFor::__for_each_n` should not instantiate the contiguous-iterator trait at all.

## Proposed fix

Nest the `if constexpr` blocks:

```cpp
if constexpr (allow_vectorization) {
  if constexpr (THRUST_NS_QUALIFIER::is_contiguous_iterator_v<RandomAccessIteratorT>) {
    // vectorized path
  }
}
```

This preserves the vectorized path when it is enabled while avoiding invalid trait instantiations when vectorization is disabled.

## Reproducer

#8778 adds two regression tests:

- A Thrust test for `thrust::lower_bound` writing to `cuda::tabulate_output_iterator`.
- A targeted CUB test for `cub::DeviceFor::ForEachN` over a `thrust::zip_iterator` containing `cuda::tabulate_output_iterator`.

--- a/cub/test/catch2_test_device_for_api.cu
+++ b/cub/test/catch2_test_device_for_api.cu
@@ -8,6 +8,10 @@
 #include <thrust/count.h>
 #include <thrust/detail/raw_pointer_cast.h>
 #include <thrust/device_vector.h>
+#include <thrust/iterator/zip_iterator.h>
+
+#include <cuda/iterator>
+#include <cuda/std/tuple>
 
 #include <c2h/catch2_test_helper.h>
 
@@ -47,6 +51,26 @@ struct odd_count_t
   }
 };
 // example-end bulk-odd-count-t
+
+struct assign_zip_value_t
+{
+  template <class Tuple>
+  __device__ void operator()(Tuple tuple) const
+  {
+    cuda::std::get<1>(tuple) = cuda::std::get<0>(tuple);
+  }
+};
+
+template <class OutputIt>
+struct tabulate_output_op
+{
+  OutputIt output;
+
+  __device__ void operator()(int idx, int value) const
+  {
+    output[idx] = value;
+  }
+};
 
 C2H_TEST("Device bulk works with temporary storage", "[bulk][device]")
 {
@@ -142,6 +166,40 @@ C2H_TEST("Device for each n works without temporary storage", "[for_each][device
   // example-end for-each-n-wo-temp-storage
 
   REQUIRE(vec == expected);
+}
+
+C2H_TEST("Device for each n works with a tabulate output iterator in a thrust zip iterator", "[for_each][device]")
+{
+  c2h::device_vector<int> input = {1, 2, 3, 4};
+  c2h::device_vector<int> output(input.size());
+
+  auto output_it = cuda::tabulate_output_iterator{tabulate_output_op<decltype(output.begin())>{output.begin()}};
+  auto zipped_it = thrust::make_zip_iterator(input.begin(), output_it);
+
+  auto result = cub::DeviceFor::ForEachN(zipped_it, input.size(), assign_zip_value_t{});
+  if (result != cudaSuccess)
+  {
+    std::cerr << "ForEachN operation failed with error code: " << result << '\n';
+  }
+
+  REQUIRE(output == input);
+}
+
+C2H_TEST("Device for each n works with a tabulate output iterator in a cuda zip iterator", "[for_each][device]")
+{
+  c2h::device_vector<int> input = {1, 2, 3, 4};
+  c2h::device_vector<int> output(input.size());
+
+  auto output_it = cuda::tabulate_output_iterator{tabulate_output_op<decltype(output.begin())>{output.begin()}};
+  auto zipped_it = cuda::make_zip_iterator(input.begin(), output_it);
+
+  auto result = cub::DeviceFor::ForEachN(zipped_it, input.size(), assign_zip_value_t{});
+  if (result != cudaSuccess)
+  {
+    std::cerr << "ForEachN operation failed with error code: " << result << '\n';
+  }
+
+  REQUIRE(output == input);
 }
 
 C2H_TEST("Device for each works with temporary storage", "[for_each][device]")

--- a/thrust/testing/catch2_test_binary_search.cu
+++ b/thrust/testing/catch2_test_binary_search.cu
@@ -3,6 +3,7 @@
 #include <thrust/sequence.h>
 #include <thrust/sort.h>
 
+#include <cuda/iterator>
 #include <cuda/std/cstdint>
 
 #include "catch2_test_helper.h"
@@ -64,6 +65,32 @@ TEST_CASE("ScalarLowerBoundDispatchImplicit", "[binary_search]")
   thrust::lower_bound(thrust::retag<my_tag>(vec.begin()), thrust::retag<my_tag>(vec.end()), 0);
 
   CHECK(13 == vec.front());
+}
+
+template <class OutputIt>
+struct tabulate_lower_bound_output_op
+{
+  OutputIt output;
+
+  _CCCL_DEVICE void operator()(int idx, int value) const
+  {
+    output[idx] = value;
+  }
+};
+
+TEST_CASE("VectorLowerBoundWithTabulateOutputIterator", "[binary_search]")
+{
+  thrust::device_vector<int> haystack = {0, 2, 4, 6};
+  thrust::device_vector<int> needles  = {1, 3, 5, 7};
+  thrust::device_vector<int> output(needles.size());
+
+  auto output_it =
+    cuda::tabulate_output_iterator{tabulate_lower_bound_output_op<decltype(output.begin())>{output.begin()}};
+
+  thrust::lower_bound(haystack.begin(), haystack.end(), needles.begin(), needles.end(), output_it);
+
+  thrust::device_vector<int> expected = {1, 2, 3, 4};
+  CHECK(output == expected);
 }
 
 TEMPLATE_LIST_TEST_CASE("ScalarUpperBoundSimple", "[binary_search]", vector_list)

--- a/thrust/testing/zip_iterator.cu
+++ b/thrust/testing/zip_iterator.cu
@@ -4,6 +4,7 @@
 #include <thrust/sequence.h>
 #include <thrust/transform.h>
 
+#include <cuda/iterator>
 #include <cuda/std/type_traits>
 
 #include <unittest/unittest.h>
@@ -15,27 +16,112 @@ void TestZipIteratorTraits()
 {
   using base_it = thrust::host_vector<int>::iterator;
 
-  using it        = thrust::zip_iterator<cuda::std::tuple<base_it, base_it>>;
-  using traits    = cuda::std::iterator_traits<it>;
-  using reference = thrust::detail::tuple_of_iterator_references<int&, int&>;
+  {
+    using it        = thrust::zip_iterator<cuda::std::tuple<base_it, base_it>>;
+    using traits    = cuda::std::iterator_traits<it>;
+    using reference = thrust::detail::tuple_of_iterator_references<int&, int&>;
 
-  static_assert(cuda::std::is_same_v<traits::difference_type, ptrdiff_t>);
-  static_assert(cuda::std::is_same_v<traits::value_type, cuda::std::tuple<int, int>>);
-  static_assert(cuda::std::is_same_v<traits::pointer, void>);
+    static_assert(cuda::std::is_same_v<traits::difference_type, ptrdiff_t>);
+    static_assert(cuda::std::is_same_v<traits::value_type, cuda::std::tuple<int, int>>);
+    static_assert(cuda::std::is_same_v<traits::pointer, void>);
 
-  static_assert(cuda::std::is_same_v<traits::reference, reference>);
-  static_assert(cuda::std::is_same_v<traits::iterator_category, ::cuda::std::random_access_iterator_tag>);
+    static_assert(cuda::std::is_same_v<traits::reference, reference>);
+    static_assert(cuda::std::is_same_v<traits::iterator_category, ::cuda::std::random_access_iterator_tag>);
 
-  static_assert(cuda::std::is_same_v<thrust::iterator_traversal_t<it>, thrust::random_access_traversal_tag>);
+    static_assert(cuda::std::is_same_v<thrust::iterator_traversal_t<it>, thrust::random_access_traversal_tag>);
 
-  static_assert(cuda::std::__has_random_access_traversal<it>);
+    static_assert(cuda::std::__has_random_access_traversal<it>);
 
-  static_assert(!cuda::std::output_iterator<it, int>);
-  static_assert(cuda::std::input_iterator<it>);
-  static_assert(cuda::std::forward_iterator<it>);
-  static_assert(cuda::std::bidirectional_iterator<it>);
-  static_assert(cuda::std::random_access_iterator<it>);
-  static_assert(!cuda::std::contiguous_iterator<it>);
+    static_assert(!cuda::std::output_iterator<it, int>);
+    static_assert(cuda::std::input_iterator<it>);
+    static_assert(cuda::std::forward_iterator<it>);
+    static_assert(cuda::std::bidirectional_iterator<it>);
+    static_assert(cuda::std::random_access_iterator<it>);
+    static_assert(!cuda::std::contiguous_iterator<it>);
+  }
+
+  { // working with proxy iterator cuda::discard_iterator
+    using it         = thrust::zip_iterator<cuda::std::tuple<base_it, cuda::discard_iterator>>;
+    using traits     = cuda::std::iterator_traits<it>;
+    using value_type = cuda::std::tuple<int, cuda::discard_iterator::__discard_proxy>;
+    using reference  = thrust::detail::tuple_of_iterator_references<int&, cuda::discard_iterator::__discard_proxy>;
+
+    static_assert(cuda::std::is_same_v<typename traits::iterator_category, cuda::std::random_access_iterator_tag>);
+    static_assert(cuda::std::is_same_v<typename traits::difference_type, cuda::std::ptrdiff_t>);
+    static_assert(cuda::std::is_same_v<typename traits::value_type, value_type>);
+    static_assert(cuda::std::is_same_v<typename traits::reference, reference>);
+
+    static_assert(!cuda::std::output_iterator<it, int>);
+    static_assert(cuda::std::input_iterator<it>);
+    static_assert(cuda::std::forward_iterator<it>);
+    static_assert(cuda::std::bidirectional_iterator<it>);
+    static_assert(cuda::std::random_access_iterator<it>);
+    static_assert(!cuda::std::contiguous_iterator<it>);
+  }
+
+  { // working with proxy iterator cuda::tabulate_output_iterator
+    using it =
+      thrust::zip_iterator<cuda::std::tuple<base_it, cuda::tabulate_output_iterator<cuda::std::plus<>, short>>>;
+    using traits     = cuda::std::iterator_traits<it>;
+    using value_type = cuda::std::tuple<int, cuda::__tabulate_proxy<cuda::std::plus<>, short>>;
+    using reference =
+      thrust::detail::tuple_of_iterator_references<int&, cuda::__tabulate_proxy<cuda::std::plus<>, short>>;
+
+    static_assert(cuda::std::is_same_v<typename traits::iterator_category, cuda::std::random_access_iterator_tag>);
+    static_assert(cuda::std::is_same_v<typename traits::difference_type, cuda::std::ptrdiff_t>);
+    static_assert(cuda::std::is_same_v<typename traits::value_type, value_type>);
+    static_assert(cuda::std::is_same_v<typename traits::reference, reference>);
+
+    static_assert(!cuda::std::output_iterator<it, int>);
+    static_assert(cuda::std::input_iterator<it>);
+    static_assert(cuda::std::forward_iterator<it>);
+    static_assert(cuda::std::bidirectional_iterator<it>);
+    static_assert(cuda::std::random_access_iterator<it>);
+    static_assert(!cuda::std::contiguous_iterator<it>);
+  }
+
+  { // working with proxy iterator cuda::transform_output_iterator
+    using it =
+      thrust::zip_iterator<cuda::std::tuple<base_it, cuda::transform_output_iterator<cuda::std::plus<>, short*>>>;
+    using traits     = cuda::std::iterator_traits<it>;
+    using value_type = cuda::std::tuple<int, cuda::__transform_output_proxy<cuda::std::plus<>, short*>>;
+    using reference =
+      thrust::detail::tuple_of_iterator_references<int&, cuda::__transform_output_proxy<cuda::std::plus<>, short*>>;
+
+    static_assert(cuda::std::is_same_v<typename traits::iterator_category, cuda::std::random_access_iterator_tag>);
+    static_assert(cuda::std::is_same_v<typename traits::difference_type, cuda::std::ptrdiff_t>);
+    static_assert(cuda::std::is_same_v<typename traits::value_type, value_type>);
+    static_assert(cuda::std::is_same_v<typename traits::reference, reference>);
+
+    static_assert(!cuda::std::output_iterator<it, int>);
+    static_assert(cuda::std::input_iterator<it>);
+    static_assert(cuda::std::forward_iterator<it>);
+    static_assert(cuda::std::bidirectional_iterator<it>);
+    static_assert(cuda::std::random_access_iterator<it>);
+    static_assert(!cuda::std::contiguous_iterator<it>);
+  }
+
+  { // working with proxy iterator cuda::transform_input_output_iterator
+    using it = thrust::zip_iterator<
+      cuda::std::tuple<base_it, cuda::transform_input_output_iterator<cuda::std::negate<>, cuda::std::plus<>, short*>>>;
+    using traits     = cuda::std::iterator_traits<it>;
+    using value_type = cuda::std::tuple<int, int>;
+    using reference  = thrust::detail::tuple_of_iterator_references<
+       int&,
+       cuda::__transform_input_output_proxy<cuda::std::negate<>, cuda::std::plus<>, short*>>;
+
+    static_assert(cuda::std::is_same_v<typename traits::iterator_category, cuda::std::random_access_iterator_tag>);
+    static_assert(cuda::std::is_same_v<typename traits::difference_type, cuda::std::ptrdiff_t>);
+    static_assert(cuda::std::is_same_v<typename traits::value_type, value_type>);
+    static_assert(cuda::std::is_same_v<typename traits::reference, reference>);
+
+    static_assert(!cuda::std::output_iterator<it, int>);
+    static_assert(cuda::std::input_iterator<it>);
+    static_assert(cuda::std::forward_iterator<it>);
+    static_assert(cuda::std::bidirectional_iterator<it>);
+    static_assert(cuda::std::random_access_iterator<it>);
+    static_assert(!cuda::std::contiguous_iterator<it>);
+  }
 }
 DECLARE_UNITTEST(TestZipIteratorTraits);
 

--- a/thrust/thrust/iterator/zip_iterator.h
+++ b/thrust/thrust/iterator/zip_iterator.h
@@ -31,6 +31,7 @@
 #include <thrust/iterator/iterator_facade.h>
 #include <thrust/iterator/iterator_traits.h>
 
+#include <cuda/__iterator/zip_common.h>
 #include <cuda/std/__iterator/advance.h>
 #include <cuda/std/__iterator/distance.h>
 #include <cuda/std/__type_traits/conditional.h>
@@ -58,18 +59,11 @@ struct make_zip_iterator_base
 template <typename... Its>
 struct make_zip_iterator_base<::cuda::std::tuple<Its...>>
 {
-  // We need this to make proxy iterators work because those have a void reference type
-  template <class Iter>
-  using zip_iterator_reference_t =
-    ::cuda::std::conditional_t<::cuda::std::is_same_v<it_reference_t<Iter>, void>,
-                               decltype(*::cuda::std::declval<Iter>()),
-                               it_reference_t<Iter>>;
-
   // reference type is the type of the tuple obtained from the iterator's reference types.
-  using reference = tuple_of_iterator_references<zip_iterator_reference_t<Its>...>;
+  using reference = tuple_of_iterator_references<::cuda::__zip_maybe_proxy_reference_t<Its>...>;
 
   // Boost's Value type is the same as reference type. using value_type = reference;
-  using value_type = ::cuda::std::tuple<it_value_t<Its>...>;
+  using value_type = ::cuda::std::tuple<::cuda::__zip_maybe_proxy_value_type_t<Its>...>;
 
   // Difference type is the first iterator's difference type
   using difference_type = it_difference_t<::cuda::std::tuple_element_t<0, ::cuda::std::tuple<Its...>>>;

